src/azure-account.ts

```diff
@@ -783,81 +802,68 @@ async function getEnvironments(): Promise<AzureAccountEnvironment[]> {
 		}
 	}
 
+	const result = [...staticEnvironments]; // make a clone
+
 	const config = workspace.getConfiguration('azure');
-	const ppe = config.get<AzureAccountEnvironment>('ppe');
+	const ppe = config.get<Environment>('ppe');
 	if (ppe) {
-		return await getPpeEnvironments(ppe, config);
+		result.push({
+			...ppe,
+			name: azurePPE,
+			validateAuthority: getValidateAuthority(ppe.activeDirectoryEndpointUrl)
+		});
 	}
-	return staticEnvironments;
-}
 
-async function getPpeEnvironments(ppe: AzureAccountEnvironment, config: WorkspaceConfiguration): Promise<AzureAccountEnvironment[]> {
-	if (!ppe.activeDirectoryEndpointUrl) {
-		return [
-			...staticEnvironments,
-			{
-				...ppe,
-				name: azurePPE
-			}
-		];
+	const customCloudEnvironment: Environment | undefined = await getCustomCloudEnvironment(config, includePartial);
+	if (customCloudEnvironment) {
+		result.push(customCloudEnvironment);
 	}
-	
-	// get api profile from user setting, this needs to be true for running azure stack
-	const apiProfile = config.get<boolean>('target_azurestack_api_profile');
-	// get validateAuthority from activeDirectoryUrl from user setting, it should be set to false only under ADFS environemnt.
-	const activeDirectoryUrl = ppe.activeDirectoryEndpointUrl.endsWith('/') ? ppe.activeDirectoryEndpointUrl.slice(0,-1) : ppe.activeDirectoryEndpointUrl;
-	const validateAuthority = activeDirectoryUrl.endsWith('/adfs') ? false : true;
-	if (apiProfile) {
-		return await getAzureStackEnvironments(ppe, validateAuthority)
-	} else {
-		return [
-			...staticEnvironments,
-			{
-				...ppe,
-				name: azurePPE,
-				validateAuthority: validateAuthority
+
+	return result;
+}
+
+async function getCustomCloudEnvironment(config: WorkspaceConfiguration, includePartial: boolean): Promise<Environment | undefined> {
```

Comment:
includePartial - need a better naming here

Reply:
we may not need this extra boolean, you can return the azurestack specific endpoints, if the cloud is set as AzureCustomCloud

Reply:
I don't think we can remove it, if we remove it we need to have another flag. I believe for this boolean, Eric wants to get all possible environment for the window picker including custom cloud. But he doesn't want to fetch endpoints every time to avoid the traffic. And we cannot just get it from user setting. If so, then users will miss the option of custom cloud in picker. But Eric wants user can use Custom Cloud even through they haven't set cloud as CustomCloud in user setting when they pick Custom cloud in window picker. As follows, https://github.com/microsoft/vscode-azure-account/blob/0411207adf53abb5baca06044fb412a48c2293ac/src/azure-account.ts#L381-L387 Probably we can give it a better name than remove this flag. What do you think about includeCustomCloud or something?